File: apps/RelayiumKit/Sources/RelayiumAppKit/DeviceInbox/InboxDestinationPlan.swift
import Foundation
@preconcurrency import RelayiumKit

/// Turning a decrypted, sender-controlled manifest into the exact set of local
/// paths one delivery is allowed to create.
///
/// AEAD proves the manifest was built by whoever holds the content key. It does
/// NOT make the names inside it safe: the sender is another machine, possibly
/// compromised, and a name is an INSTRUCTION to this filesystem. Everything below
/// treats a manifest name as hostile input.
///
/// The plan is computed ONCE, before anything is downloaded, and journalled
/// before anything is committed. That ordering is what makes a crash resumable:
/// the set of destinations a task may ever create is fixed and durable before the
/// first one exists, so recovery never re-derives it from a directory that has
/// since changed — which would walk the collision suffix forward and deliver the
/// same file twice.
///
/// This is deliberately NOT `ManifestWriter`/`safePathSegments`. Those two
/// STRIP control and bidi characters and then write straight into the
/// destination, cleaning up on failure; that is right for an interactive
/// download the user is watching, where a repaired name beats a refused
/// transfer. A Device Inbox delivery is unattended: a repaired name is a file the
/// user never approved appearing on their disk under a name nobody chose, and a
/// direct write has no crash boundary a journal can answer. Reusing the writer
/// here would import its ownership and cleanup semantics along with its path
/// rules.

/// One manifest entry bound to the one path it may ever create.
///
/// `name` and `destination` are plaintext-derived and therefore LOCAL ONLY: they
/// live in the journal so a crash can resume, and they are never logged, never
/// printed, and never sent to central.
public struct InboxPlanEntry: Codable, Equatable, Sendable {
    public let index: Int
    public let name: String
    public let size: Int
    /// Absolute path, as an on-disk string rather than a `URL`, because it is
    /// journalled and compared byte-for-byte across runs.
    public let destination: String

    public init(index: Int, name: String, size: Int, destination: String) {
        self.index = index
        self.name = name
        self.size = size
        self.destination = destination
    }
}

public enum InboxPlanError: Error, Equatable, Sendable {
    /// A manifest entry this device refuses to materialise. Terminal: the same
    /// bytes are refused the same way on every retry.
    case unsafeName(index: Int)
    /// Two entries resolve to one destination, including differing only by case.
    /// A REFUSAL rather than a rename: two entries that differ only by case are
    /// the sender describing two files, and quietly renaming one would hide that
    /// this receiver cannot represent what was sent.
    case duplicateDestination(index: Int)
    /// Every deterministic candidate name is taken. A human has to look.
    case noFreeName(index: Int)
}

public enum InboxDestinationPlan {
    /// Bounds directory nesting created for one task. Deep trees are legitimate;
    /// unbounded depth is a cheap way to exhaust path limits and make cleanup
    /// expensive.
    static let maxPathDepth = 32
    /// Bounds the deterministic `name (2)` search. Reaching it means the folder
    /// genuinely holds thousands of same-named files, which is a human problem.
    static let maxCollisionIndex = 1000

    /// The per-task staging area, created INSIDE the receive directory so the
    /// final commit is a same-filesystem `link` and can be atomic. A staging area
    /// on another filesystem would silently degrade the commit to a copy,
    /// reintroducing the partially-written-file window this design removes.
    public static let stagingDirectoryName = ".relayium-incoming"

    /// The entries in the receive directory that belong to this component, and
    /// which a manifest may therefore never name.
    ///
    /// Both are real hazards, not tidiness. A delivery into `stagingDirectoryName`
    /// would land ON its own staged source, so the commit would link the file to
    /// itself and then unlink it — reporting `saved` with nothing on disk. A
    /// delivery named `probeName` would be DELETED by the next writability probe,
    /// which removes a stale probe file it assumes it left behind.
    static var reservedTopLevelNames: Set<String> {
        [stagingDirectoryName, InboxReceiveFolder.probeName]
    }

    /// Windows device names. A file cannot be created with any of them, with or
    /// without an extension, in any case. Refused on macOS TOO, on purpose: one
    /// manifest is then accepted or refused identically on every receiver, and a
    /// name only some of a user's devices can receive is worse than one none can.
    static let reservedDeviceNames: Set<String> = [
        "con", "prn", "aux", "nul",
        "com1", "com2", "com3", "com4", "com5", "com6", "com7", "com8", "com9",
        "lpt1", "lpt2", "lpt3", "lpt4", "lpt5", "lpt6", "lpt7", "lpt8", "lpt9",
    ]

    /// Validate one manifest name and return it as a cleaned, slash-separated
    /// relative path.
    ///
    /// The rules, and what each one stops:
    ///
    ///   - empty / oversized / invalid UTF-8: unrepresentable or unbounded input.
    ///     (Swift `String` is already valid UTF-8 by construction; the JSON
    ///     decoder that produced it would have refused otherwise.)
    ///   - control scalars, including NUL and DEL: a name that can truncate a C
    ///     string or rewrite a terminal line. REFUSED, not stripped.
    ///   - bidi controls: a name that reads one way in a UI and another on disk.
    ///   - a leading `/` or a drive/UNC prefix: an ABSOLUTE destination, which
    ///     would leave the receive directory entirely.
    ///   - `\` anywhere: the separator is `/` by protocol. A backslash is a legal
    ///     byte in a POSIX name but means "directory separator" on Windows, so one
    ///     manifest would produce different trees on different receivers.
    ///   - `.` or `..` components: traversal.
    ///   - empty components (`a//b`): two spellings of one path.
    ///   - a component ending in `.` or a space: Windows strips both silently, so
    ///     `x ` and `x` would collide there and not here.
    ///   - a Windows reserved device name.
    /// Returns nil for a name this device refuses. Nil rather than a thrown
    /// error because the caller is the only thing that knows which manifest INDEX
    /// is being refused, and an error carrying a placeholder index would be a
    /// value the caller has to remember to overwrite.
    public static func checkedRelativePath(_ name: String) -> String? {
        guard !name.isEmpty, name.utf8.count <= MANIFEST_MAX_NAME_BYTES else { return nil }
        for scalar in name.unicodeScalars {
            let v = scalar.value
            if v <= 0x1F || (v >= 0x7F && v <= 0x9F) { return nil }
        }
        guard stripBidi(name) == name else { return nil }
        guard !name.contains("\\") else { return nil }
        guard !name.hasPrefix("/") else { return nil }
        // `C:foo` is drive-relative and `C:/foo` drive-absolute on Windows.
        let chars = Array(name)
        if chars.count >= 2, chars[1] == ":" { return nil }
        let parts = name.split(separator: "/", omittingEmptySubsequences: false).map(String.init)
        guard parts.count <= maxPathDepth else { return nil }
        for part in parts {
            guard !part.isEmpty, part != ".", part != ".." else { return nil }
            guard !part.hasSuffix("."), !part.hasSuffix(" ") else { return nil }
            let stem = part.split(separator: ".", maxSplits: 1,
                                  omittingEmptySubsequences: false)[0].lowercased()
            guard !reservedDeviceNames.contains(stem) else { return nil }
        }
        return name
    }

    /// Split a file name into the part a collision suffix is inserted after and
    /// the extension that must be preserved.
    ///
    /// Preserving the extension is a product requirement, not cosmetics:
    /// `report (2)` with `.pdf` moved or dropped stops opening in the right
    /// application. The three shapes that need care:
    ///
    ///   - `notes`         -> (`notes`, ``)          no extension to keep
    ///   - `.bashrc`       -> (`.bashrc`, ``)        a dotfile is all stem
    ///   - `backup.tar.gz` -> (`backup`, `.tar.gz`)  a compound archive suffix
    static func splitBase(_ base: String) -> (stem: String, ext: String) {
        for compound in [".tar.gz", ".tar.bz2", ".tar.xz", ".tar.zst", ".tar.lz4"] {
            if base.count > compound.count, base.lowercased().hasSuffix(compound) {
                let cut = base.index(base.endIndex, offsetBy: -compound.count)
                return (String(base[..<cut]), String(base[cut...]))
            }
        }
        // A leading dot is part of the name, not an extension marker.
        guard let dot = base.lastIndex(of: "."), dot != base.startIndex,
              base.index(after: dot) != base.endIndex else {
            return (base, "")
        }
        return (String(base[..<dot]), String(base[dot...]))
    }

    /// The deterministic safe rename (PRD §9): `name (2)`, then `name (3)`, with
    /// the extension preserved.
    static func collisionName(_ base: String, _ n: Int) -> String {
        let (stem, ext) = splitBase(base)
        return "\(stem) (\(n))\(ext)"
    }

    /// Resolve every manifest entry to the exact path it may create, refusing the
    /// whole task rather than partially planning it.
    ///
    /// `exists` is injected so the plan can be computed against a snapshot and so
    /// a test can plant a destination between planning and commit. It must report
    /// whether ANYTHING occupies that path — a file, a directory, a socket, a FIFO,
    /// a device node or a DANGLING symlink — because all of those occupy the name
    /// and none may be written through.
    ///
    /// Ordering is manifest order and the collision search is lowest-index-first,
    /// so the same manifest against the same directory always produces the same
    /// plan. That determinism is what lets a resumed task compare its journalled
    /// plan against reality instead of guessing.
    /// Takes v2 manifest ITEMS rather than the shared manifest's file entries.
    /// A `.text` item has no name and no destination, so it is refused here as
    /// well as gated by the receiver: a planner that assumed its caller had
    /// already separated the kinds would create a destination for a nameless
    /// item the first time one forgot.
    public static func plan(root: URL, files: [InboxManifestItem],
                            exists: (String) -> Bool = InboxDestinationPlan.pathExists)
        throws -> [InboxPlanEntry] {
        let absRoot = root.standardizedFileURL.path
        var plan: [InboxPlanEntry] = []
        // Every path this plan will create, so two entries cannot be given the
        // same destination even when the directory is empty.
        var taken = Set<String>()
        // The same set, case-folded: a case-insensitive filesystem (APFS, HFS+)
        // would collapse two differently-cased names into one file.
        var lowered = Set<String>()
        // The case-folded destinations the MANIFEST asked for, before any
        // collision suffix.
        //
        // Kept separately from `lowered`, and the distinction is the whole reason
        // both exist. Two manifest entries whose REQUESTED destinations differ
        // only by case are the sender describing two files this receiver cannot
        // tell apart, and that is refused — renaming one would hide it. But an
        // entry whose requested name happens to equal the SUFFIX an earlier entry
        // was given (`a.txt` was taken, so it became `a (2).txt`, and the manifest
        // also contains a real `a (2).txt`) describes two perfectly distinct
        // files; that one steps aside to the next free name, exactly as it would
        // against any other occupied name.
        var requested = Set<String>()

        func occupied(_ path: String) -> Bool {
            taken.contains(path) || lowered.contains(path.lowercased()) || exists(path)
        }

        for (i, file) in files.enumerated() {
            guard file.kind == .file, let name = file.name,
                  let relative = checkedRelativePath(name) else {
                throw InboxPlanError.unsafeName(index: i)
            }
            let top = relative.split(separator: "/", maxSplits: 1).first.map(String.init) ?? relative
            guard !reservedTopLevelNames.contains(top) else {
                throw InboxPlanError.unsafeName(index: i)
            }

            var destination = absRoot
            for component in relative.split(separator: "/") {
                destination += "/" + component
            }
            // Belt and braces over the component checks: the joined path must
            // still be under the root. Cheap, and it catches any future
            // `sanitize` regression before it reaches the filesystem.
            guard destination.hasPrefix(absRoot + "/") else {
                throw InboxPlanError.unsafeName(index: i)
            }
            guard requested.insert(destination.lowercased()).inserted else {
                throw InboxPlanError.duplicateDestination(index: i)
            }

            var final = destination
            if occupied(final) {
                var found = false
                guard let slash = destination.lastIndex(of: "/") else {
                    throw InboxPlanError.unsafeName(index: i)
                }
                let directory = String(destination[..<slash])
                let base = String(destination[destination.index(after: slash)...])
                for n in 2...maxCollisionIndex {
                    let candidate = directory + "/" + collisionName(base, n)
                    if !occupied(candidate) {
                        final = candidate
                        found = true
                        break
                    }
                }
                guard found else { throw InboxPlanError.noFreeName(index: i) }
            }
            taken.insert(final)
            lowered.insert(final.lowercased())
            plan.append(InboxPlanEntry(index: i, name: name, size: file.size,
                                       destination: final))
        }
        return plan
    }

    /// Whether anything at all occupies `path`, without following a final
    /// symlink.
    ///
    /// `lstat` rather than `stat` is the point: a dangling symlink is not a free
    /// name, and a symlink pointing at somebody's `~/.ssh/authorized_keys` is
    /// emphatically not a free name.
    public static func pathExists(_ path: String) -> Bool {
        var st = stat()
        return lstat(path, &st) == 0
    }
}

File: apps/android/app/src/main/kotlin/com/relayium/android/inbox/InboxDestinationPlan.kt
package com.relayium.android.inbox

import com.relayium.protocol.inbox.InboxManifest
import com.relayium.protocol.inbox.InboxManifestItem
import com.relayium.protocol.inbox.InboxManifestKind
import java.io.File
import java.util.Locale

/**
 * Turning a decrypted, sender-controlled manifest into the exact set of local
 * paths one delivery is allowed to create.
 *
 * AEAD proves the manifest was written by whoever holds the content key. It does
 * NOT make the names inside it safe: the sender is another machine, possibly
 * compromised, and a name is an INSTRUCTION to this filesystem. Everything below
 * treats a manifest name as hostile input.
 *
 * The plan is computed ONCE, before anything is downloaded, and journalled
 * before anything is committed. That ordering is what makes a crash resumable:
 * the set of destinations a task may ever create is fixed and durable before the
 * first one exists, so recovery never re-derives it from a directory that has
 * since changed — which would walk the collision suffix forward and deliver the
 * same file twice.
 *
 * ## Why names are REFUSED rather than repaired
 *
 * The app's interactive download path sanitises a name and writes it, which is
 * right for a transfer a person is watching: a repaired name beats a refused
 * one. A Device Inbox delivery is unattended, so a repaired name is a file the
 * user never approved appearing under a name nobody chose. Every rule below
 * therefore refuses.
 */

/** One manifest entry bound to the one path it may ever create.
 *
 *  [name] and [destination] are plaintext-derived and therefore LOCAL ONLY: they
 *  live in the journal so a crash can resume, and never reach central or a log. */
data class InboxPlanEntry(
    val index: Int,
    val name: String,
    val size: Long,
    /** Absolute path as a STRING, because it is journalled and compared
     *  byte-for-byte across runs. */
    val destination: String,
)

/** Why a manifest cannot be materialised here. Carries the INDEX, never the name. */
class InboxPlanException(val reason: Reason, val index: Int) :
    RuntimeException("relayium inbox plan: $reason at $index") {

    enum class Reason {
        /** A name this device refuses to create. Terminal: the same bytes are
         *  refused the same way on every attempt. */
        UNSAFE_NAME,

        /**
         * Two entries resolve to one destination, including differing only by
         * case.
         *
         * A REFUSAL rather than a rename: two entries that differ only by case
         * are the sender describing two files, and quietly renaming one would
         * hide that this receiver cannot represent what was sent.
         */
        DUPLICATE_DESTINATION,

        /** Every deterministic candidate name is taken. A person has to look. */
        NO_FREE_NAME,

        /** The manifest describes a message, which has no destination at all. */
        NOT_A_FILE_MANIFEST,
    }
}

object InboxDestinationPlan {

    /** Bounds directory nesting created for one task. Deep trees are legitimate;
     *  unbounded depth is a cheap way to exhaust path limits. */
    const val MAX_PATH_DEPTH = 32

    /** Bounds the deterministic `name (2)` search for a free DIRECTORY name. */
    const val MAX_COLLISION_INDEX = 1000

    /**
     * Windows device names, refused on Android TOO.
     *
     * Deliberate: one manifest is then accepted or refused identically on every
     * receiver, and a name only some of a user's devices can receive is worse
     * than one none can.
     */
    private val RESERVED_DEVICE_NAMES = setOf(
        "con", "prn", "aux", "nul",
        "com1", "com2", "com3", "com4", "com5", "com6", "com7", "com8", "com9",
        "lpt1", "lpt2", "lpt3", "lpt4", "lpt5", "lpt6", "lpt7", "lpt8", "lpt9",
    )

    /**
     * Names this component owns. A manifest may never produce one, at the top
     * level of a delivery or as the root of a path inside it.
     *
     * Each would do something specific and bad:
     *
     *  * [InboxCommit.RECEIPT_NAME] is written into the staged tree AFTER every
     *    file has been written and digested. A sender that named a file
     *    `.relayium-delivery` would have it silently replaced by the receipt —
     *    a delivery that downloaded and verified correctly and arrived
     *    corrupted. Refused before a byte moves.
     *  * [InboxContainer.STAGING] is where the delivery is assembled;
     *  * [InboxContainer.PROBE] is deleted by the next readiness check.
     */
    private val RESERVED_TOP_LEVEL = setOf(
        InboxCommit.RECEIPT_NAME,
        InboxContainer.STAGING,
        InboxContainer.PROBE,
    )

    /**
     * Validate one manifest name and return it as a cleaned, slash-separated
     * relative path, or null if this device refuses it.
     *
     * Null rather than a thrown error because only the caller knows which
     * manifest INDEX is being refused, and an error carrying a placeholder index
     * is a value someone has to remember to overwrite.
     *
     * The rules, and what each one stops:
     *
     *  * empty or oversized — unrepresentable or unbounded input;
     *  * control scalars including NUL and DEL — a name that can truncate a C
     *    string or rewrite a terminal line. Refused, not stripped;
     *  * bidi controls — a name that reads one way in a UI and another on disk;
     *  * a leading `/` or a drive/UNC prefix — an ABSOLUTE destination, which
     *    would leave the container entirely;
     *  * `\` anywhere — the separator is `/` by protocol, and a backslash is a
     *    legal byte in a POSIX name but a separator on Windows, so one manifest
     *    would produce different trees on different receivers;
     *  * `.` or `..` components — traversal;
     *  * empty components (`a//b`) — two spellings of one path;
     *  * a component ending in `.` or a space — Windows strips both silently, so
     *    `x ` and `x` would collide there and not here;
     *  * a Windows reserved device name;
     *  * a top-level name this component owns.
     */
    fun checkedRelativePath(name: String): String? {
        if (name.isEmpty() || name.toByteArray(Charsets.UTF_8).size > InboxManifest.MAX_NAME_BYTES) {
            return null
        }
        for (ch in name) {
            val code = ch.code
            if (code <= 0x1F || (code in 0x7F..0x9F)) return null
            if (code in BIDI_CONTROLS) return null
        }
        if (name.contains('\\')) return null
        if (name.startsWith('/')) return null
        // `C:foo` is drive-relative and `C:/foo` drive-absolute on Windows.
        if (name.length >= 2 && name[1] == ':') return null

        val parts = name.split('/')
        if (parts.size > MAX_PATH_DEPTH) return null
        for (part in parts) {
            if (part.isEmpty() || part == "." || part == "..") return null
            if (part.endsWith('.') || part.endsWith(' ')) return null
            val stem = part.substringBefore('.').lowercase(Locale.ROOT)
            if (stem in RESERVED_DEVICE_NAMES) return null
        }
        if (parts.first() in RESERVED_TOP_LEVEL) return null
        return name
    }

    /**
     * Choose the directory this delivery will publish into.
     *
     * Picked ONCE, journalled before anything is created, and never recomputed:
     * a name derived again later — against a container that now holds this
     * task's own output — would walk the suffix forward and publish the same
     * delivery twice.
     *
     * A whole directory per delivery, rather than files dropped into the
     * container root, is what makes publication a single atomic rename of
     * something this task exclusively built. It also means the container root is
     * never a place this feature creates, truncates or deletes a file the user
     * might own.
     */
    fun planTaskDirectory(root: File, items: List<InboxManifestItem>, taskId: String): File {
        val base = baseName(items)
        // The directory name comes from the sender as well, so it cannot be one
        // of this component's own entries either.
        if (base in RESERVED_TOP_LEVEL) return File(root, DEFAULT_DIRECTORY_NAME)
        for (n in 1..MAX_COLLISION_INDEX) {
            val candidate = if (n == 1) base else "$base ($n)"
            val directory = File(root, candidate)
            if (!directory.exists()) return directory
        }
        // Deterministic and certainly free: the task id is unique per delivery.
        return File(root, "$base-${InboxId.checked(taskId, "taskId")}")
    }

    /**
     * A readable folder name for a delivery.
     *
     * Derived from what was actually sent, so the user sees something meaningful
     * rather than an opaque id — and passed through the same refusal every other
     * name gets, because it comes from the sender too.
     */
    private fun baseName(items: List<InboxManifestItem>): String {
        val first = items.firstOrNull()?.name?.substringBefore('/').orEmpty()
        val stem = first.substringBeforeLast('.', first)
        val checked = if (stem.isEmpty()) null else checkedRelativePath(stem)
        // A single component only: a folder name is one name, and a nested one
        // would put the delivery somewhere the plan did not describe.
        return checked?.takeIf { !it.contains('/') } ?: DEFAULT_DIRECTORY_NAME
    }

    /** The fallback when a manifest offers nothing usable as a folder name. */
    const val DEFAULT_DIRECTORY_NAME = "Delivery"

    /**
     * The complete, ordered set of destinations one delivery may create, INSIDE
     * its own published directory.
     *
     * Because that directory is fresh and exclusively this task's, there is no
     * collision with the user's own files to step around: the only clash that
     * can occur is two manifest entries naming one path, and that is a refusal
     * rather than a rename — two entries differing only by case are the sender
     * describing two files, and quietly renaming one would hide that this
     * receiver cannot represent what was sent.
     */
    fun plan(taskDirectory: File, items: List<InboxManifestItem>): List<InboxPlanEntry> {
        val taken = HashSet<String>()
        val entries = ArrayList<InboxPlanEntry>(items.size)
        for ((index, item) in items.withIndex()) {
            val name = item.name
            if (item.kind != InboxManifestKind.FILE || name == null) {
                throw InboxPlanException(InboxPlanException.Reason.NOT_A_FILE_MANIFEST, index)
            }
            val relative = checkedRelativePath(name)
                ?: throw InboxPlanException(InboxPlanException.Reason.UNSAFE_NAME, index)
            val destination = File(taskDirectory, relative).path
            if (!taken.add(destination.lowercase(Locale.ROOT))) {
                throw InboxPlanException(InboxPlanException.Reason.DUPLICATE_DESTINATION, index)
            }
            entries.add(InboxPlanEntry(index, relative, item.size, destination))
        }
        return entries
    }

    /** U+200E..U+200F and U+202A..U+202E and U+2066..U+2069. */
    private val BIDI_CONTROLS: Set<Int> =
        (0x200E..0x200F).toSet() + (0x202A..0x202E).toSet() + (0x2066..0x2069).toSet()
}
